File: backend-v2/routers/clients.py
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from sqlalchemy import or_, func
from datetime import datetime, timezone

from database import get_db
from models import Client, User, Appointment
from schemas import ClientCreate, ClientUpdate, Client as ClientSchema, ClientList, ClientHistory
from utils.auth import get_current_user
from services import client_service

router = APIRouter(prefix="/clients", tags=["clients"])


@router.post("/", response_model=ClientSchema)
async def create_client(
    client_data: ClientCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a new client"""
    # Check if client with email already exists
    existing_client = db.query(Client).filter(Client.email == client_data.email).first()
    if existing_client:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Client with this email already exists"
        )
    
    try:
        new_client = client_service.create_client(
            db=db,
            client_data=client_data.model_dump(),
            created_by_id=current_user.id
        )
        return new_client
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.get("/", response_model=ClientList)
async def list_clients(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    search: Optional[str] = None,
    customer_type: Optional[str] = None,
    tags: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List clients with pagination and filtering"""
    query = db.query(Client)
    
    # Search by name, email, or phone
    if search:
        search_filter = or_(
            Client.first_name.ilike(f"%{search}%"),
            Client.last_name.ilike(f"%{search}%"),
            Client.email.ilike(f"%{search}%"),
            Client.phone.ilike(f"%{search}%")
        )
        query = query.filter(search_filter)
    
    # Filter by customer type
    if customer_type:
        query = query.filter(Client.customer_type == customer_type)
    
    # Filter by tags
    if tags:
        query = query.filter(Client.tags.ilike(f"%{tags}%"))
    
    # Get total count
    total = query.count()
    
    # Apply pagination
    offset = (page - 1) * page_size
    clients = query.offset(offset).limit(page_size).all()
    
    return {
        "clients": clients,
        "total": total,
        "page": page,
        "page_size": page_size
    }


@router.get("/{client_id}", response_model=ClientSchema)
async def get_client(
    client_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get a specific client by ID"""
    client = db.query(Client).filter(Client.id == client_id).first()
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found"
        )
    return client


@router.put("/{client_id}", response_model=ClientSchema)
async def update_client(
    client_id: int,
    client_update: ClientUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update client information"""
    client = db.query(Client).filter(Client.id == client_id).first()
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found"
        )
    
    # Update only provided fields
    update_data = client_update.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(client, field, value)
    
    # Update the updated_at timestamp
    client.updated_at = datetime.now(timezone.utc)
    
    db.commit()
    db.refresh(client)
    
    return client


@router.delete("/{client_id}")
async def delete_client(
    client_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Delete a client (admin only)"""
    if current_user.role != "admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only admins can delete clients"
        )
    
    client = db.query(Client).filter(Client.id == client_id).first()
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found"
        )
    
    db.delete(client)
    db.commit()
    
    return {"message": "Client deleted successfully"}


@router.get("/{client_id}/history", response_model=ClientHistory)
async def get_client_history(
    client_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get client appointment history and statistics"""
    client = db.query(Client).filter(Client.id == client_id).first()
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found"
        )
    
    try:
        # Update client metrics automatically
        client_service.update_client_metrics(db, client_id)
        
        # Get all appointments for this client
        appointments = db.query(Appointment).filter(
            Appointment.client_id == client_id
        ).order_by(Appointment.start_time.desc()).all()
        
        # Get analytics
        analytics = client_service.get_client_analytics(db, client_id)
        
        return {
            "appointments": appointments,
            "total_appointments": analytics["total_scheduled"],
            "total_spent": analytics["total_spent"],
            "average_ticket": analytics["average_ticket"],
            "no_shows": analytics["no_show_count"],
            "cancellations": analytics["cancellation_count"]
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.put("/{client_id}/customer-type")
async def update_customer_type(
    client_id: int,
    customer_type: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update customer type (new, returning, vip, at_risk)"""
    if customer_type not in ["new", "returning", "vip", "at_risk"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid customer type"
        )
    
    client = db.query(Client).filter(Client.id == client_id).first()
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found"
        )
    
    client.customer_type = customer_type
    client.updated_at = datetime.now(timezone.utc)
    
    db.commit()
    
    return {"message": f"Customer type updated to {customer_type}"}


@router.post("/search")
async def search_clients(
    query: str,
    limit: int = 10,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Quick search for clients (for autocomplete)"""
    try:
        clients = client_service.search_clients_advanced(
            db=db,
            query=query,
            limit=limit
        )
        
        # Return simplified data for autocomplete
        return [
            {
                "id": client.id,
                "name": f"{client.first_name} {client.last_name}",
                "email": client.email,
                "phone": client.phone,
                "customer_type": client.customer_type
            }
            for client in clients
        ]
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.get("/{client_id}/analytics")
async def get_client_analytics(
    client_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get comprehensive analytics for a client"""
    try:
        analytics = client_service.get_client_analytics(db, client_id)
        return analytics
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.get("/{client_id}/recommendations")
async def get_client_recommendations(
    client_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get personalized recommendations for a client"""
    try:
        recommendations = client_service.get_client_recommendations(db, client_id)
        return recommendations
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.get("/{client_id}/communication-preferences")
async def get_client_communication_preferences(
    client_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get client communication preferences"""
    try:
        preferences = client_service.get_client_communication_preferences(db, client_id)
        return preferences
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.put("/{client_id}/communication-preferences")
async def update_client_communication_preferences(
    client_id: int,
    preferences: dict,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update client communication preferences"""
    try:
        updated_client = client_service.update_client_communication_preferences(
            db, client_id, preferences
        )
        return {"message": "Communication preferences updated successfully"}
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.post("/{client_id}/notes")
async def add_client_note(
    client_id: int,
    note_data: dict,  # {"note": str, "note_type": str}
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Add a note to a client's record"""
    try:
        note_result = client_service.add_client_note(
            db=db,
            client_id=client_id,
            note=note_data.get("note", ""),
            added_by_id=current_user.id,
            note_type=note_data.get("note_type", "general")
        )
        return note_result
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.put("/{client_id}/tags")
async def update_client_tags(
    client_id: int,
    tag_data: dict,  # {"tags": List[str]}
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update client tags"""
    try:
        updated_client = client_service.update_client_tags(
            db, client_id, tag_data.get("tags", [])
        )
        return {"message": "Tags updated successfully", "tags": updated_client.tags}
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.get("/dashboard/metrics")
async def get_client_dashboard_metrics(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get dashboard metrics for client management"""
    try:
        metrics = client_service.get_client_dashboard_metrics(db)
        return metrics
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.get("/advanced-search")
async def advanced_client_search(
    query: Optional[str] = Query(None),
    customer_type: Optional[str] = Query(None),
    tags: Optional[str] = Query(None),
    min_visits: Optional[int] = Query(None),
    max_days_since_visit: Optional[int] = Query(None),
    limit: int = Query(50, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Advanced client search with multiple filters"""
    try:
        clients = client_service.search_clients_advanced(
            db=db,
            query=query,
            customer_type=customer_type,
            tags=tags,
            min_visits=min_visits,
            max_days_since_visit=max_days_since_visit,
            limit=limit
        )
        return clients
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )

File: backend-v2/frontend-v2/lib/api/clients.ts
/**
 * Clients API Client
 * 
 * Provides comprehensive client/customer management functionality including:
 * - Client profile creation and management
 * - Client search and filtering
 * - Client history and analytics
 * - Communication preferences and notes
 * - Client recommendations and insights
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'

// ===============================
// TypeScript Interfaces
// ===============================

export interface ClientCreate {
  first_name: string
  last_name: string
  email: string
  phone?: string
  date_of_birth?: string // YYYY-MM-DD format
  notes?: string
  tags?: string
  preferred_barber_id?: number
  preferred_services?: string[]
  communication_preferences?: Record<string, any>
}

export interface ClientUpdate {
  first_name?: string
  last_name?: string
  email?: string
  phone?: string
  date_of_birth?: string
  notes?: string
  tags?: string
  preferred_barber_id?: number
  preferred_services?: string[]
  communication_preferences?: Record<string, any>
}

export interface Client {
  id: number
  first_name: string
  last_name: string
  email: string
  phone?: string
  date_of_birth?: string
  notes?: string
  tags?: string
  preferred_barber_id?: number
  preferred_services?: string[]
  communication_preferences?: Record<string, any>
  customer_type: string
  total_visits: number
  total_spent: number
  average_ticket: number
  visit_frequency_days?: number
  no_show_count: number
  cancellation_count: number
  referral_count: number
  first_visit_date?: string
  last_visit_date?: string
  created_at: string
  updated_at: string
  is_active: boolean
}

export interface ClientListResponse {
  clients: Client[]
  total: number
  page: number
  page_size: number
}

export interface ClientHistory {
  appointments: Appointment[]
  total_appointments: number
  total_spent: number
  average_ticket: number
  no_shows: number
  cancellations: number
}

export interface ClientAnalytics {
  client_lifetime_value: number
  predicted_next_visit: string
  visit_frequency_trend: string
  spending_trend: string
  risk_score: number
  satisfaction_score?: number
  retention_probability: number
  recommended_services: string[]
}

export interface ClientRecommendations {
  services: string[]
  upsell_opportunities: string[]
  retention_strategies: string[]
  next_appointment_suggestions: string[]
}

export interface ClientCommunicationPreferences {
  email_notifications: boolean
  sms_notifications: boolean
  marketing_emails: boolean
  appointment_reminders: boolean
  promotional_offers: boolean
  preferred_contact_method: string
  notification_timing: Record<string, any>
}

export interface ClientNote {
  id: number
  client_id: number
  note: string
  created_by: string
  created_at: string
  is_private: boolean
}

export interface ClientNoteCreate {
  note: string
  is_private: boolean
}

export interface ClientTagsUpdate {
  tags: string[]
}

export interface ClientSearchRequest {
  query: string
  filters?: {
    customer_type?: string
    tags?: string[]
    preferred_barber_id?: number
    date_range?: {
      start: string
      end: string
    }
  }
}

export interface ClientSearchResponse {
  clients: Client[]
  total: number
  query: string
}

export interface ClientDashboardMetrics {
  total_clients: number
  new_clients_this_month: number
  active_clients: number
  retention_rate: number
  average_lifetime_value: number
  top_customer_types: Array<{
    type: string
    count: number
    percentage: number
  }>
  recent_client_activity: Array<{
    client_id: number
    client_name: string
    activity_type: string
    activity_date: string
  }>
}

export interface AdvancedSearchFilters {
  query?: string
  customer_type?: string
  tags?: string[]
  preferred_barber_id?: number
  min_total_spent?: number
  max_total_spent?: number
  min_visits?: number
  max_visits?: number
  first_visit_start?: string
  first_visit_end?: string
  last_visit_start?: string
  last_visit_end?: string
  has_no_shows?: boolean
  has_cancellations?: boolean
  risk_level?: string
}

export interface ClientFilters {
  page?: number
  page_size?: number
  search?: string
  customer_type?: string
  tags?: string
}

// ===============================
// Utility Functions
// ===============================

/**
 * Get authorization headers with current JWT token
 */
function getAuthHeaders(): Record<string, string> {
  const token = localStorage.getItem('access_token')
  return {
    'Content-Type': 'application/json',
    ...(token && { Authorization: `Bearer ${token}` })
  }
}

/**
 * Handle API response and extract data
 */
async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const errorData = await response.json().catch(() => ({ detail: 'Unknown error' }))
    throw new Error(errorData.detail || `Request failed with status ${response.status}`)
  }
  return response.json()
}

/**
 * Build query string from parameters
 */
function buildQueryString(params: Record<string, any>): string {
  const filtered = Object.entries(params)
    .filter(([_, value]) => value !== undefined && value !== null && value !== '')
    .map(([key, value]) => {
      if (Array.isArray(value)) {
        return `${encodeURIComponent(key)}=${value.map(v => encodeURIComponent(v)).join(',')}`
      }
      return `${encodeURIComponent(key)}=${encodeURIComponent(value)}`
    })
  
  return filtered.length > 0 ? `?${filtered.join('&')}` : ''
}

// ===============================
// Clients API Client
// ===============================

export const clientsApi = {
  /**
   * Create a new client
   */
  async createClient(clientData: ClientCreate): Promise<Client> {
    const response = await fetch(`${API_BASE_URL}/api/v1/clients/`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify(clientData)
    })

    return handleResponse(response)
  },

  /**
   * Get list of clients with pagination and filtering
   */
  async getClients(filters: ClientFilters = {}): Promise<ClientListResponse> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${buildQueryString(filters)}`,
      {
        headers: getAuthHeaders()
      }
    )

    return handleResponse(response)
  },

  /**
   * Get specific client by ID
   */
  async getClient(clientId: number): Promise<Client> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}`,
      {
        headers: getAuthHeaders()
      }
    )

    return handleResponse(response)
  },

  /**
   * Update client information
   */
  async updateClient(clientId: number, updates: ClientUpdate): Promise<Client> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}`,
      {
        method: 'PUT',
        headers: getAuthHeaders(),
        body: JSON.stringify(updates)
      }
    )

    return handleResponse(response)
  },

  /**
   * Delete client
   */
  async deleteClient(clientId: number): Promise<{ message: string }> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}`,
      {
        method: 'DELETE',
        headers: getAuthHeaders()
      }
    )

    return handleResponse(response)
  },

  /**
   * Get client appointment history and metrics
   */
  async getClientHistory(clientId: number): Promise<ClientHistory> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}/history`,
      {
        headers: getAuthHeaders()
      }
    )

    return handleResponse(response)
  },

  /**
   * Update client customer type
   */
  async updateClientCustomerType(
    clientId: number, 
    customerType: string
  ): Promise<{ message: string }> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}/customer-type`,
      {
        method: 'PUT',
        headers: getAuthHeaders(),
        body: JSON.stringify({ customer_type: customerType })
      }
    )

    return handleResponse(response)
  },

  /**
   * Search clients by query
   */
  async searchClients(searchRequest: ClientSearchRequest): Promise<ClientSearchResponse> {
    const response = await fetch(`${API_BASE_URL}/api/v1/clients/search`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify(searchRequest)
    })

    return handleResponse(response)
  },

  /**
   * Advanced client search with multiple filters
   */
  async advancedSearchClients(filters: AdvancedSearchFilters): Promise<ClientListResponse> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/advanced-search${buildQueryString(filters)}`,
      {
        headers: getAuthHeaders()
      }
    )

    return handleResponse(response)
  },

  /**
   * Get client analytics and insights
   */
  async getClientAnalytics(clientId: number): Promise<ClientAnalytics> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}/analytics`,
      {
        headers: getAuthHeaders()
      }
    )

    return handleResponse(response)
  },

  /**
   * Get client recommendations
   */
  async getClientRecommendations(clientId: number): Promise<ClientRecommendations> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}/recommendations`,
      {
        headers: getAuthHeaders()
      }
    )

    return handleResponse(response)
  },

  /**
   * Get client communication preferences
   */
  async getClientCommunicationPreferences(clientId: number): Promise<ClientCommunicationPreferences> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}/communication-preferences`,
      {
        headers: getAuthHeaders()
      }
    )

    return handleResponse(response)
  },

  /**
   * Update client communication preferences
   */
  async updateClientCommunicationPreferences(
    clientId: number,
    preferences: Partial<ClientCommunicationPreferences>
  ): Promise<ClientCommunicationPreferences> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}/communication-preferences`,
      {
        method: 'PUT',
        headers: getAuthHeaders(),
        body: JSON.stringify(preferences)
      }
    )

    return handleResponse(response)
  },

  /**
   * Add a note to client profile
   */
  async addClientNote(clientId: number, noteData: ClientNoteCreate): Promise<ClientNote> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}/notes`,
      {
        method: 'POST',
        headers: getAuthHeaders(),
        body: JSON.stringify(noteData)
      }
    )

    return handleResponse(response)
  },

  /**
   * Update client tags
   */
  async updateClientTags(clientId: number, tags: string[]): Promise<Client> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/${clientId}/tags`,
      {
        method: 'PUT',
        headers: getAuthHeaders(),
        body: JSON.stringify({ tags })
      }
    )

    return handleResponse(response)
  },

  /**
   * Get client dashboard metrics
   */
  async getClientDashboardMetrics(): Promise<ClientDashboardMetrics> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/clients/dashboard/metrics`,
      {
        headers: getAuthHeaders()
      }
    )

    return handleResponse(response)
  },

  // ===============================
  // Utility Methods
  // ===============================

  /**
   * Get client full name
   */
  getFullName(client: Client): string {
    return `${client.first_name} ${client.last_name}`
  },

  /**
   * Format client phone number
   */
  formatPhone(phone?: string): string {
    if (!phone) return ''
    
    // Remove all non-digits
    const digits = phone.replace(/\D/g, '')
    
    // Format as (XXX) XXX-XXXX for US numbers
    if (digits.length === 10) {
      return `(${digits.slice(0, 3)}) ${digits.slice(3, 6)}-${digits.slice(6)}`
    }
    
    return phone
  },

  /**
   * Calculate client age from date of birth
   */
  calculateAge(dateOfBirth?: string): number | null {
    if (!dateOfBirth) return null
    
    const birth = new Date(dateOfBirth)
    const today = new Date()
    let age = today.getFullYear() - birth.getFullYear()
    const monthDiff = today.getMonth() - birth.getMonth()
    
    if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birth.getDate())) {
      age--
    }
    
    return age
  },

  /**
   * Get client status based on recent activity
   */
  getClientStatus(client: Client): 'active' | 'inactive' | 'new' | 'at_risk' {
    const now = new Date()
    const lastVisit = client.last_visit_date ? new Date(client.last_visit_date) : null
    const firstVisit = client.first_visit_date ? new Date(client.first_visit_date) : null
    
    // New client (first visit within 30 days)
    if (firstVisit && (now.getTime() - firstVisit.getTime()) < (30 * 24 * 60 * 60 * 1000)) {
      return 'new'
    }
    
    // At risk (no visit in 90+ days)
    if (lastVisit && (now.getTime() - lastVisit.getTime()) > (90 * 24 * 60 * 60 * 1000)) {
      return 'at_risk'
    }
    
    // Active (visit within 60 days)
    if (lastVisit && (now.getTime() - lastVisit.getTime()) < (60 * 24 * 60 * 60 * 1000)) {
      return 'active'
    }
    
    return 'inactive'
  },

  /**
   * Get client loyalty tier based on spending/visits
   */
  getLoyaltyTier(client: Client): 'bronze' | 'silver' | 'gold' | 'platinum' {
    if (client.total_spent >= 1000 || client.total_visits >= 20) {
      return 'platinum'
    } else if (client.total_spent >= 500 || client.total_visits >= 10) {
      return 'gold'
    } else if (client.total_spent >= 200 || client.total_visits >= 5) {
      return 'silver'
    }
    return 'bronze'
  },

  /**
   * Parse client tags into array
   */
  parseTags(tags?: string): string[] {
    if (!tags) return []
    return tags.split(',').map(tag => tag.trim()).filter(tag => tag.length > 0)
  },

  /**
   * Format tags array into string
   */
  formatTags(tags: string[]): string {
    return tags.join(', ')
  },

  /**
   * Check if client has specific tag
   */
  hasTag(client: Client, tag: string): boolean {
    const clientTags = this.parseTags(client.tags)
    return clientTags.some(t => t.toLowerCase() === tag.toLowerCase())
  },

  /**
   * Get next suggested appointment date based on visit frequency
   */
  getNextSuggestedAppointment(client: Client): Date | null {
    if (!client.last_visit_date || !client.visit_frequency_days) {
      return null
    }
    
    const lastVisit = new Date(client.last_visit_date)
    const nextDate = new Date(lastVisit)
    nextDate.setDate(nextDate.getDate() + client.visit_frequency_days)
    
    return nextDate
  }
}

export default clientsApi